main/lsp/updates.cc

```diff
@@ -265,37 +266,29 @@ LSPLoop::TypecheckRun LSPLoop::tryFastPath(unique_ptr<core::GlobalState> gs,
                 }
                 globalStateHashes[fref.id()] = hashes[i];
             }
-            fast_sort(changedHashes);
-            changedHashes.resize(
-                std::distance(changedHashes.begin(), std::unique(changedHashes.begin(), changedHashes.end())));
+            core::NameHash::sortAndDedupe(changedHashes);
         }
     }
 
     if (takeFastPath) {
         Timer timeit(logger, "fast_path");
-        if (allFiles) {
-            subset.clear();
-            for (int i = 1; i < finalGs->filesUsed(); i++) {
-                core::FileRef fref(i);
-                if (fref.data(*finalGs).sourceType == core::File::Type::Normal) {
-                    subset.emplace_back(core::FileRef(i));
-                }
-            }
-        } else {
-            int i = -1;
-            for (auto &oldHash : globalStateHashes) {
-                i++;
-                vector<core::NameHash> intersection;
-                std::set_intersection(changedHashes.begin(), changedHashes.end(), oldHash.usages.usages.begin(),
-                                      oldHash.usages.usages.end(), std::back_inserter(intersection));
-                if (!intersection.empty()) {
-                    auto ref = core::FileRef(i);
-                    logger->debug("Added {} to update set as used a changed method",
-                                  !ref.exists() ? "" : ref.data(*finalGs).path());
-                    subset.emplace_back(ref);
-                }
+        int i = -1;
+        for (auto &oldHash : globalStateHashes) {
+            i++;
+            vector<core::NameHash> intersection;
+            std::set_intersection(changedHashes.begin(), changedHashes.end(), oldHash.usages.sends.begin(),
+                                  oldHash.usages.sends.end(), std::back_inserter(intersection));
+            if (!intersection.empty()) {
+                auto ref = core::FileRef(i);
+                logger->debug("Added {} to update set as used a changed method",
+                              !ref.exists() ? "" : ref.data(*finalGs).path());
+                subset.emplace_back(ref);
             }
         }
+        // Remove any duplicate files.
+        fast_sort(subset);
```

Comment:
Before this dedupe logic, I think we might have been typechecking `changedFiles` twice...

Reply:
Huh, are you sure? won't we report errors twice than? Why wasn't this found by tests?

Reply:
Tests suppress duplicate errors. (This was the behavior in `test_corpus` before I refactored it, and it was decided back then to keep the behavior.)

You can see the issue here:

https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20%23%20Change%20'String'%20to%20'Integer'%0A%20%20sig%20%7Bparams(x%3A%20Integer).returns(String)%7D%0A%20%20def%20bar(x)%0A%20%20%20%20x.to_s%0A%20%20end%0Aend%0A%0Aj%20%3D%20T.let(A.new.bar(10)%2C%20String)%0A%0A%0A

Reply:
> won't we report errors twice then?

tl;dr: Yes; we do report errors twice.

Reply:
Duplicate error ignore logic is here:

https://github.com/sorbet/sorbet/blob/master/test/helpers/position_assertions.cc#L628